# 3. Coursera - Using Python to Access Web Data

In [2]:
# Search for lines that contain 'From'
import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    if re.search('From:', line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


In [3]:
# Search for lines that start with 'From'
import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    if re.search('^From:', line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


In [4]:
# Search for lines that start with 'F', followed by
# 2 characters, followed by 'm:'
import re
hand =  open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    if re.search('^F..m:', line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


In [5]:
# Search for lines that start with From and have an at sign
import re
hand = open('mbox-short.txt')
for line  in hand:
    line =  line.rstrip()
    if re.search('^From:.+@', line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


In [1]:
import re
s = 'A message from csev@umich.edu to cwen@iupui.edu about meeting @2PM'
lst = re.findall('\S+@\S+', s)
print(lst)

['csev@umich.edu', 'cwen@iupui.edu']


In [2]:
# Search for lines that have an at sign between characters
import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('\S+@\S+', line)
    if len(x) > 0:
        print(x)

['stephen.marquard@uct.ac.za']
['<postmaster@collab.sakaiproject.org>']
['<200801051412.m05ECIaH010327@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['apache@localhost)']
['source@collab.sakaiproject.org;']
['stephen.marquard@uct.ac.za']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['stephen.marquard@uct.ac.za']
['louis@media.berkeley.edu']
['<postmaster@collab.sakaiproject.org>']
['<200801042308.m04N8v6O008125@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['apache@localhost)']
['source@collab.sakaiproject.org;']
['louis@media.berkeley.edu']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['louis@media.berkeley.edu']
['zqian@umich.edu']
['<postmaster@collab.sakaiproject.org>']
['<200801042109.m04L92hb007923@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject

In [3]:
# Search for lines that have an at sign between characters
# The characters must be a letter or number
import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('[a-zA-Z0-9]\S+@\S+[a-zA-Z]', line)
    if len(x) > 0:
        print(x)

['stephen.marquard@uct.ac.za']
['postmaster@collab.sakaiproject.org']
['200801051412.m05ECIaH010327@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['apache@localhost']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['stephen.marquard@uct.ac.za']
['louis@media.berkeley.edu']
['postmaster@collab.sakaiproject.org']
['200801042308.m04N8v6O008125@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['apache@localhost']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['louis@media.berkeley.edu']
['zqian@umich.edu']
['postmaster@collab.sakaiproject.org']
['200801042109.m04L92hb007923@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject

In [1]:
# Search for lines that start with 'X' followed by any non
# whitespace characters and ':'
# followed by a space and any number.
# The number can include a decimal.
import re
hand  = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    if re.search('^X\S*: [0-9.]+', line):
        print(line)

X-DSPAM-Confidence: 0.8475
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6178
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6961
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7565
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7626
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7556
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7002
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7615
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7601
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7605
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6959
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7606
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7559
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7605
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6932
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7558
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6526
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6948
X-DSPAM-Probability: 0.0000
X-DSPAM-Co

In [6]:
# Search for lines that start with 'X' followed by any 
# non whitespace characters and ':' followed by a space 
# and any number. The number can include a decimal.
# Then print the number if it is greater than zero.
import re
hand  = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('^X\S*: ([0-9.]+)', line)
    if len(x) > 0:
        print(float(x[0]))

0.8475
0.0
0.6178
0.0
0.6961
0.0
0.7565
0.0
0.7626
0.0
0.7556
0.0
0.7002
0.0
0.7615
0.0
0.7601
0.0
0.7605
0.0
0.6959
0.0
0.7606
0.0
0.7559
0.0
0.7605
0.0
0.6932
0.0
0.7558
0.0
0.6526
0.0
0.6948
0.0
0.6528
0.0
0.7002
0.0
0.7554
0.0
0.6956
0.0
0.6959
0.0
0.7556
0.0
0.9846
0.0
0.8509
0.0
0.9907
0.0


In [8]:
# Search for lines that with 'Details: rev='
# followed by numbers and '.'
# Then print the number if its is greather than zero
import re
hand = open('mbox-short.txt')
for line in hand:
    line =  line.rstrip()
    x = re.findall('^Details:.*rev=([0-9.]+)', line)
    if len(x) > 0:
        print(int(x[0]))

39772
39771
39770
39769
39766
39765
39764
39763
39762
39761
39760
39759
39758
39757
39756
39755
39754
39753
39752
39751
39750
39749
39746
39745
39744
39743
39742


In [14]:
# Search for lines that start with From and a character
# followed by a two digit number between 00 and 99 followed by ':'
# Then print the number if it is greater than zego
import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('^From .* ([0-9][0-9]:[0-9][0-9])', line)
    if len(x) > 0:
        print(x)

['09:14']
['18:10']
['16:10']
['15:46']
['15:03']
['14:50']
['11:37']
['11:35']
['11:12']
['11:11']
['11:11']
['11:10']
['10:38']
['10:17']
['10:04']
['09:05']
['07:02']
['06:08']
['04:49']
['04:33']
['04:07']
['19:51']
['17:18']
['17:07']
['16:34']
['16:29']
['16:23']


In [18]:
import re
x = 'We just received $10.00 for cookies.'
y = re.findall('\$[0-9]+.[0-9]+', x)
y

['$10.00']

In [24]:
!grep '^From:' mbox-short.txt

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


<b>Exercise 1:</b> Write a simple program to simulate the operation of the grep command on Unix. Ask the user to enter a regular expression and count the number of lines that matched the regular expression:

$ python grep.py<br>
Enter a regular expression: ^Author<br>
mbox.txt had 1798 lines that matched ^Author

$ python grep.py<br>
Enter a regular expression: ^X-<br>
mbox.txt had 14368 lines that matched ^X-<br>

python grep.py<br>
Enter a regular expression: java<br>
mbox.txt had 4175 lines that matched java$

In [3]:
import re
try:
    fhand = open('mbox.txt')
except:
    print('Error to open a file mbox.txt')
else:
    cmd = input('Enter a regular expression: ')
    lst = list()
    for line in fhand:
        line = line.rstrip()
        x = re.findall(cmd, line)
        lst += x
    print('mbox.txt had',len(lst), 'lines that matched', cmd)
        

Enter a regular expression: java$
mbox.txt had 4218 lines that matched java$


<b>Exercise 2:</b> Write a program to look for lines of the form:

New Revision: 39772

Extract the number from each of the lines using a regular expression and the findall() method. Compute the average of the numbers and print out the average as an integer.

Enter file:mbox.txt<br>
38549

Enter file:mbox-short.txt<br>
39756

In [24]:
import re

fhand = input('Enter file: ')
try:
    fhand = open(fhand)
except:
    print('Error to open a file', fhand)
else:
    lst = list()
    for line in fhand:
        line = line.rstrip()
        x = re.findall('^New Revision:.(\d+)', line)
        if len(x) > 0:
            lst += x
    lst = [int(i) for i in lst]
    print(int(sum(lst)/len(lst)))

Enter file: mbox.txt
38549


Bienvenido Marcelo Palma Salas de Using Python to Access Web Data

Finding Numbers in a Haystack

In this assignment you will read through and parse a file with text and numbers. You will extract all the numbers in the file and compute the sum of the numbers.

<b>Data Files</b><br>
We provide two files for this assignment. One is a sample file where we give you the sum for your testing and the other is the actual data you need to process for the assignment.

Sample data: http://py4e-data.dr-chuck.net/regex_sum_42.txt (There are 90 values with a sum=445833)
Actual data: http://py4e-data.dr-chuck.net/regex_sum_373489.txt (There are 86 values and the sum ends with 494)
These links open in a new window. Make sure to save the file into the same folder as you will be writing your Python program. Note: Each student will have a distinct data file for the assignment - so only use your own data file for analysis.

<b>Data Format</b><br>
The file contains much of the text from the introduction of the textbook except that random numbers are inserted throughout the text. Here is a sample of the output you might see:

<br><br>
Why should you learn to write programs? 7746<br>
12 1929 8827<br>
Writing programs (or programming) is a very creative <br>
7 and rewarding activity.  You can write programs for <br>
many reasons, ranging from making your living to solving<br>
8837 a difficult data analysis problem to having fun to helping 128<br>
someone else solve a problem.  This book assumes that <br>
everyone needs to know how to program ...<br>
<br><br>

The sum for the sample text above is 27486. The numbers can appear anywhere in the line. There can be any number of numbers in each line (including none).

<b>Handling The Data</b><br>
The basic outline of this problem is to read the file, look for integers using the re.findall(), looking for a regular expression of '[0-9]+' and then converting the extracted strings to integers and summing up the integers.

<b>Turn in Assignent</b>

Enter the sum from the actual data and your Python code below:<br>
Sum: <br>
 (ends with 494) 
Python code:<br>

In [46]:
import re
try:
    fhand = open('regex_sum_373489.txt')
except:
    print('Error to open a file', fhand)
else:
    lst = list()
    for line in fhand:
        line = line.rstrip()
        x = re.findall('\d+', line)
        if len(x) > 0:
            lst += x
    lst = [int(i) for i in lst]
    print(sum(lst))

455494


In [15]:
import socket

mysock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
mysock.connect(('data.pr4e.org', 80))
cmd = 'GET http://data.pr4e.org/romeo.txt HTTP/1.0\r\n\r\n'.encode()
mysock.send(cmd)

while True:
    data = mysock.recv(512)
    if len(data) < 1:
        break
    print(data.decode(), end='')

mysock.close()

HTTP/1.1 200 OK
Date: Sun, 12 Apr 2020 18:19:43 GMT
Server: Apache/2.4.18 (Ubuntu)
Last-Modified: Sat, 13 May 2017 11:22:22 GMT
ETag: "a7-54f6609245537"
Accept-Ranges: bytes
Content-Length: 167
Cache-Control: max-age=0, no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: Wed, 11 Jan 1984 05:00:00 GMT
Connection: close
Content-Type: text/plain

But soft what light through yonder window breaks
It is the east and Juliet is the sun
Arise fair sun and kill the envious moon
Who is already sick and pale with grief


In [4]:
!cat romeo.txt

But soft what light through yonder window breaks
It is the east and Juliet is the sun
Arise fair sun and kill the envious moon
Who is already sick and pale with grief


In [2]:
import socket
import time

HOST = 'data.pr4e.org'
PORT = 80
mysock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
mysock.connect((HOST, PORT))
mysock.sendall(b'GET http://data.pr4e.org/cover3.jpg HTTP/1.0\r\n\r\n')
count = 0
picture = b""

while True:
    data = mysock.recv(5120)
    if len(data) < 1: break
    #time.sleep(0.25)
    count = count + len(data)
    print(len(data), count)
    picture = picture + data

mysock.close()

# Look for the end of the header (2 CRLF)
pos = picture.find(b"\r\n\r\n")
print('Header lenght', pos)
print(picture[:pos].decode())

# Skip past the header and save the picture data
picture = picture[pos+4:]
fhand = open("stuff.jpg", "wb")
fhand.write(picture)
fhand.close()

2880 2880
2880 5760
2880 8640
5120 13760
640 14400
1440 15840
5120 20960
5120 26080
5120 31200
5120 36320
5120 41440
1760 43200
1440 44640
2880 47520
5120 52640
5120 57760
5120 62880
5120 68000
5120 73120
4640 77760
1440 79200
5120 84320
5120 89440
1280 90720
5120 95840
5120 100960
5120 106080
5120 111200
5120 116320
5120 121440
5120 126560
5120 131680
5120 136800
5120 141920
5120 147040
5120 152160
5120 157280
5120 162400
1760 164160
1440 165600
5120 170720
5120 175840
5120 180960
5120 186080
5120 191200
5120 196320
5120 201440
5120 206560
5120 211680
5120 216800
5120 221920
5120 227040
3568 230608
Header lenght 394
HTTP/1.1 200 OK
Date: Wed, 15 Apr 2020 14:24:29 GMT
Server: Apache/2.4.18 (Ubuntu)
Last-Modified: Mon, 15 May 2017 12:27:40 GMT
ETag: "38342-54f8f2e5b6277"
Accept-Ranges: bytes
Content-Length: 230210
Vary: Accept-Encoding
Cache-Control: max-age=0, no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: Wed, 11 Jan 1984 05:00:00 GMT
Connection: close
Content-Type: ima

In [3]:
import urllib.request as req

fhand =  req.urlopen('http://data.pr4e.org/romeo.txt')
for line in fhand:
    print(line.decode().strip())

But soft what light through yonder window breaks
It is the east and Juliet is the sun
Arise fair sun and kill the envious moon
Who is already sick and pale with grief


In [4]:
import urllib.request as req

fhand = req.urlopen('http://data.pr4e.org/romeo.txt')

counts = dict()
for line in fhand:
    words = line.decode().split()
    for word in words:
        counts[word] = counts.get(word, 0) + 1

print(counts)

{'But': 1, 'soft': 1, 'what': 1, 'light': 1, 'through': 1, 'yonder': 1, 'window': 1, 'breaks': 1, 'It': 1, 'is': 3, 'the': 3, 'east': 1, 'and': 3, 'Juliet': 1, 'sun': 2, 'Arise': 1, 'fair': 1, 'kill': 1, 'envious': 1, 'moon': 1, 'Who': 1, 'already': 1, 'sick': 1, 'pale': 1, 'with': 1, 'grief': 1}


In [6]:
from urllib import request, parse, error

img = request.urlopen('http://data.pr4e.org/cover3.jpg').read()
fhand = open('cover3.jpg', 'wb')
fhand.write(img)
fhand.close()

In [18]:
!ls -l cover3.jpg

-rw-r--r-- 1 ecram ecram 230210 abr 15 11:44 cover3.jpg


In [19]:
from urllib import request as req, parse, error

img = req.urlopen('http://data.pr4e.org/cover3.jpg')
fhand = open('cover31.jpg', 'wb')
size = 0
while True:
    info = img.read(10000)
    if len(info) < 1:
        break
    size = size +len(info)
    fhand.write(info)

print(size, 'characters copied.')
fhand.close()

230210 characters copied.


In [20]:
ls -l cover31.jpg

-rw-r--r-- 1 ecram ecram 230210 abr 15 11:56 cover31.jpg


In [2]:
# Search for link values within URL input
from urllib import request as req, parse, error
import re
import ssl

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
html = req.urlopen(url, context=ctx).read()
links = re.findall(b'href="(http[s]?://.*?)"', html)
for link in links:
    print(link.decode())

Enter - https://www.lasca.ic.unicamp.br/site/
http://www.unicamp.br/unicamp/
http://www.unicamp.br/unicamp/
http://www.ic.unicamp.br/
http://www.ic.unicamp.br/pos
http://www.ic.unicamp.br/
http://www.lasca.ic.unicamp.br/pub/ubuntu-releases/
http://www.lasca.ic.unicamp.br/pub/ubuntu-releases/
http://www.ic.unicamp.br/
http://www.unicamp.br/unicamp/
http://www.unicamp.br/unicamp/
http://www.ic.unicamp.br/
http://www.ic.unicamp.br/pos
http://www.ic.unicamp.br/
http://www.lasca.ic.unicamp.br/pub/ubuntu-releases/
http://www.lasca.ic.unicamp.br/pub
http://www.prefeitura.unicamp.br/servicos/diretoria-de-servicos-de-transporte
http://www.emdec.com.br/ABusInf/index.asp?MODULO=&OpcPPD=0&EdMaxPe=600&LtOpcTipoDia=0&LtOpcPrioriz=O&LtOpcHorario=19&LIGACAO=&PontosNotOrig=59&PontosNotDest=18&TIPO_LOCAL_ORIG=63&TIPO_LOCAL_DEST=16&CdNotIDDestino=154&CdNotIDOrigem=&LOCAL_DEST=&NumDestino=&CruzDestino=&LOCAL_ORIG=&BAIRRO_DEST=
http://www.ic.unicamp.br/~dfaranha
http://www.ic.unicamp.br/node/204
http://www

Instalar beautifulsoup

pip install beautifulsoup4

In [4]:
# To run this, diwnload the BeautifulSoup zip file
# http://www.py4e.com/code3/bs4.zip
# and unzip it in the same directory ad this file

from urllib import request as req, parse, error
from bs4 import BeautifulSoup as bs
import ssl

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
html = req.urlopen(url,context=ctx).read()
soup = bs(html, 'html.parser')

# Retraive all of the anchor tags
tags = soup('a')
for tag in tags:
    print(tag.get('href', None))

Enter - https://docs.python.org
genindex.html
py-modindex.html
https://www.python.org/
#
whatsnew/3.8.html
whatsnew/index.html
tutorial/index.html
library/index.html
reference/index.html
using/index.html
howto/index.html
installing/index.html
distributing/index.html
extending/index.html
c-api/index.html
faq/index.html
py-modindex.html
genindex.html
glossary.html
search.html
contents.html
bugs.html
https://devguide.python.org/docquality/#helping-with-documentation
about.html
license.html
copyright.html
download.html
https://docs.python.org/3.9/
https://docs.python.org/3.8/
https://docs.python.org/3.7/
https://docs.python.org/3.6/
https://docs.python.org/3.5/
https://docs.python.org/2.7/
https://www.python.org/doc/versions/
https://www.python.org/dev/peps/
https://wiki.python.org/moin/BeginnersGuide
https://wiki.python.org/moin/PythonBooks
https://www.python.org/doc/av/
https://devguide.python.org/
genindex.html
py-modindex.html
https://www.python.org/
#
copyright.html
https://www.python

In [5]:
print(soup)


<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta charset="utf-8"/><title>3.8.2 Documentation</title>
<link href="_static/pydoctheme.css" rel="stylesheet" type="text/css"/>
<link href="_static/pygments.css" rel="stylesheet" type="text/css"/>
<script data-url_root="./" id="documentation_options" src="_static/documentation_options.js" type="text/javascript"></script>
<script src="_static/jquery.js" type="text/javascript"></script>
<script src="_static/underscore.js" type="text/javascript"></script>
<script src="_static/doctools.js" type="text/javascript"></script>
<script src="_static/language_data.js" type="text/javascript"></script>
<script src="_static/sidebar.js" type="text/javascript"></script>
<link href="_static/opensearch.xml" rel="search" title="Search within Python 3.8.2 documentation" type="application/opensearchdescription+xml"/>
<link href="about.html" rel="author" title="About these documents"/>
<link href="genindex.html" rel="index" title="Index"/>

In [42]:
# To run this, download the BeautifulSoup zip file
# http://www.py4e.com/code3/bs4.zip
# and unzip it in the same directory as this file

from urllib import request as req, parse, error
from bs4 import BeautifulSoup as bs
import ssl

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
html = req.urlopen(url, context=ctx).read()
soup = bs(html, "html.parser")

# Retrieve all of the anchot tags
tags = soup('a')
for tag in tags:
    # Look at the parts of a tag
    print('TAG:', tag)
    print('URL:', tag.get('href', None))
    print('Contents:', tag.contents[0])
    print('Attrs:', tag.attrs)

Enter - http://www.hackerav.com/?post=242539
TAG: <a href="http://www.hackerav.com/">码农人生		</a>
URL: http://www.hackerav.com/
Contents: 码农人生		
Attrs: {'href': 'http://www.hackerav.com/'}
TAG: <a href="https://weibo.com/1896610970" target="_blank"><i class="iconfont icon-weibo"></i></a>
URL: https://weibo.com/1896610970
Contents: <i class="iconfont icon-weibo"></i>
Attrs: {'href': 'https://weibo.com/1896610970', 'target': '_blank'}
TAG: <a href="http://www.hackerav.com/">首页</a>
URL: http://www.hackerav.com/
Contents: 首页
Attrs: {'href': 'http://www.hackerav.com/'}
TAG: <a href="http://www.hackerav.com/?sort=4">疑难问题</a>
URL: http://www.hackerav.com/?sort=4
Contents: 疑难问题
Attrs: {'href': 'http://www.hackerav.com/?sort=4'}
TAG: <a href="http://www.hackerav.com/t">微语</a>
URL: http://www.hackerav.com/t
Contents: 微语
Attrs: {'href': 'http://www.hackerav.com/t'}
TAG: <a href="http://www.hackerav.com/?sort=5">软件使用</a>
URL: http://www.hackerav.com/?sort=5
Contents: 软件使用
Attrs: {'href': 'http://www

IndexError: list index out of range

In [9]:
!curl -O http://www.py4e.com/cover.jpg

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0


In [10]:
!ls -l cover.jpg

-rw-r--r-- 1 ecram ecram 0 abr 15 17:46 cover.jpg


In [11]:
!wget http://www.py4e.com/cover.jpg

--2020-04-15 17:49:04--  http://www.py4e.com/cover.jpg
Resolving www.py4e.com (www.py4e.com)... 2606:4700:3032::681b:9ea6, 2606:4700:3034::681b:9fa6, 104.27.159.166, ...
Connecting to www.py4e.com (www.py4e.com)|2606:4700:3032::681b:9ea6|:80... failed: Connection timed out.
Connecting to www.py4e.com (www.py4e.com)|2606:4700:3034::681b:9fa6|:80... ^C


In [1]:
import os
import urllib.request, urllib.parse, urllib.error

print('Please enter a URL like http://data.pr4e.org/cover3.jpg')
urlstr = input().strip()
img = urllib.request.urlopen(urlstr)

# Get the last "word"
words = urlstr.split('/')
fname = words[-1]

# Don't overwrite the file
if os.path.exists(fname):
    if input('Replace ' + fname + ' (Y/n)?') != 'Y':
        print('Data not copied')
        exit()
    print('Replacing', fname)

fhand = open(fname, 'wb')
size = 0
while True:
    info = img.read(100000)
    if len(info) < 1: break
    size = size + len(info)
    fhand.write(info)

print(size, 'characters copied to', fname)
fhand.close()

Please enter a URL like http://data.pr4e.org/cover3.jpg
http://data.pr4e.org/cover3.jpg
Replace cover3.jpg (Y/n)?Y
Replacing cover3.jpg
230210 characters copied to cover3.jpg


In [13]:
import requests as req
from bs4 import BeautifulSoup as bs
import ssl

url = input('Enter - ')
html = req.get(url, verify=False)
soup = bs(html.text, "html.parser")

# Retrieve all of the anchor tags
tags = soup('a')
for tag in tags:
    # Look at the parts of a tag
    print('TAG:', tag)
    print('URL:', tag.get('href', None))
    print('Contents:', tag.contents[0])
    print('Attrs:', tag.attrs)


Enter - http://www.dr-chuck.com/page1.htm
TAG: <a href="http://www.dr-chuck.com/page2.htm">
Second Page</a>
URL: http://www.dr-chuck.com/page2.htm
Contents: 
Second Page
Attrs: {'href': 'http://www.dr-chuck.com/page2.htm'}


<b>Exercise 1:</b> Change the socket program socket1.py to prompt the user for the URL so it can read any web page. You can use split('/') to break the URL into its component parts so you can extract the host name for the socket connect call. Add error checking using try and except to handle the condition where the user enters an improperly formatted or non-existent URL.

In [20]:
import socket
import re

#ejemplo = 'http://data.pr4e.org/romeo.txt'

mysock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
web = input('Enter - ')

url = re.findall('[http[s]?://]?(.*?)/', web)

try:
    print(url)
    mysock.connect((url[0], 80))
except:
    print('Error URL format.')
else:
    cmd = ('GET '+web+' HTTP/1.0\r\n\r\n').encode()
    mysock.send(cmd)

    while True:
        data = mysock.recv(512)
        if len(data) < 1:
            break
        print(data.decode(),end='')

    mysock.close()

Enter - http://data.pr4e.org/romeo.txt
['data.pr4e.org']
HTTP/1.1 200 OK
Date: Fri, 17 Apr 2020 20:54:51 GMT
Server: Apache/2.4.18 (Ubuntu)
Last-Modified: Sat, 13 May 2017 11:22:22 GMT
ETag: "a7-54f6609245537"
Accept-Ranges: bytes
Content-Length: 167
Cache-Control: max-age=0, no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: Wed, 11 Jan 1984 05:00:00 GMT
Connection: close
Content-Type: text/plain

But soft what light through yonder window breaks
It is the east and Juliet is the sun
Arise fair sun and kill the envious moon
Who is already sick and pale with grief


<b>Exercise 2:</b> Change your socket program so that it counts the number of characters it has received and stops displaying any text after it has shown 3000 characters. The program should retrieve the entire document and count the total number of characters and display the count of the number of characters at the end of the document.

In [98]:
import socket
import re

#ejemplo = 'http://data.pr4e.org/romeo.txt'

mysock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
web = input('Enter - ')

url = re.findall('[http[s]?://]?(.*?)/', web)

try:
    mysock.connect((url[0], 80))
except:
    print('Error URL format.')
else:
    cmd = ('GET '+web+' HTTP/1.0\r\n\r\n').encode()
    mysock.send(cmd)
    text = ''.encode()
    while True:
        data = mysock.recv(512)
        if len(data) < 1:
            break
        text = text + data
    salida = text.decode()
    print(salida[:3000],end='')
    print('\nNumber of Characters: ', len(salida))
    mysock.close()

Enter - http://data.pr4e.org/intro-short.txt
HTTP/1.1 200 OK
Date: Sat, 18 Apr 2020 20:20:49 GMT
Server: Apache/2.4.18 (Ubuntu)
Last-Modified: Sat, 13 May 2017 11:22:22 GMT
ETag: "1d3-54f6609240717"
Accept-Ranges: bytes
Content-Length: 467
Cache-Control: max-age=0, no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: Wed, 11 Jan 1984 05:00:00 GMT
Connection: close
Content-Type: text/plain

Why should you learn to write programs?

Writing programs (or programming) is a very creative 
and rewarding activity.  You can write programs for 
many reasons, ranging from making your living to solving
a difficult data analysis problem to having fun to helping
someone else solve a problem.  This book assumes that 
everyone needs to know how to program, and that once 
you know how to program you will figure out what you want 
to do with your newfound skills.  

Number of Characters:  837


<b>Exercise 3:</b> Use urllib to replicate the previous exercise of (1) retrieving the document from a URL, (2) displaying up to 3000 characters, and (3) counting the overall number of characters in the document. Don’t worry about the headers for this exercise, simply show the first 3000 characters of the document contents.

In [25]:
import urllib.request as req
import re

# ejemplo = 'http://data.pr4e.org/romeo.txt'
http = input('Enter - ')

try:
    fhand =  req.urlopen(http)
except:
    print('Error URL format.')
else:
    text = ''.encode()
    for line in fhand:
        text +=line
print(text.decode().strip()[:3000])
print('\nNumber of Characters: ', len(text.decode().strip()))

Enter - http://data.pr4e.org/romeo-full.txt
Romeo and Juliet
Act 2, Scene 2 

SCENE II. Capulet's orchard.

Enter ROMEO

ROMEO

He jests at scars that never felt a wound.
JULIET appears above at a window

But, soft! what light through yonder window breaks?
It is the east, and Juliet is the sun.
Arise, fair sun, and kill the envious moon,
Who is already sick and pale with grief,
That thou her maid art far more fair than she:
Be not her maid, since she is envious;
Her vestal livery is but sick and green
And none but fools do wear it; cast it off.
It is my lady, O, it is my love!
O, that she knew she were!
She speaks yet she says nothing: what of that?
Her eye discourses; I will answer it.
I am too bold, 'tis not to me she speaks:
Two of the fairest stars in all the heaven,
Having some business, do entreat her eyes
To twinkle in their spheres till they return.
What if her eyes were there, they in her head?
The brightness of her cheek would shame those stars,
As daylight doth a lamp; her e

<b>Exercise 4:</b> Change the urllinks.py program to extract and count paragraph (p) tags from the retrieved HTML document and display the count of the paragraphs as the output of your program. Do not display the paragraph text, only count them. Test your program on several small web pages as well as some larger web pages.

In [1]:
%%time
from urllib.request import Request, urlopen
import urllib.parse, urllib.error
from bs4 import BeautifulSoup
import ssl

# ejemplo = 'http://www.dr-chuck.com/page1.htm'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
req = Request(url, headers={'User-Agent': 'Mozilla/5.0'})
with urlopen(req) as response:
    html = response.read()
    soup = BeautifulSoup(html, 'html.parser')
    # Retrieve and count all of the paragraphs tags
    paragraphs = soup('p')
    print(len(paragraphs))

Enter - https://stackoverflow.com/questions/28670835/python-socket-client-post-parameters
29
CPU times: user 746 ms, sys: 30.2 ms, total: 776 ms
Wall time: 2.74 s


<b>Exercise 5:</b> (Advanced) Change the socket program so that it only shows data after the headers and a blank line have been received. Remember that recv receives characters (newlines and all), not lines.

In [1]:
%%time
import socket
import re

#ejemplo = 'http://www.students.ic.unicamp.br/~ra089053/'
mysock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
web = input('Enter - ')
url = re.findall('[http[s]?://]?(.*?)/', web)

try:
    #print(url[0])
    mysock.connect((url[0], 80))
except:
    print('Error URL format.')
else:
    cmd = ('GET '+web+' HTTP/1.0\r\n\r\n').encode()
    mysock.send(cmd)
    text = ''.encode()
    while True:
        data = mysock.recv(512)
        if not data: break
        text += data
    data = text.decode()
    #print(data[:200],end='')
    position=data.find('\r\n\r\n')
    print(data[position+4:])
    mysock.close()

Enter - http://www.students.ic.unicamp.br/~ra089053/
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<title>Marcelo Invert Palma Salas</title>
<link href="css/estilo.css" rel="stylesheet" type="text/css" />
<style type="text/css">
body {
	background-color: #FFFFFF;
}
a:link {
	color: #000000;
}
a:visited {
	color: #000000;
}
</style>
</head>

<body text="#000000" link="#000000">

<div class="container">
  <div class="sidebar1">
    <ul class="nav">
      <li></li>
      <li></li>
      <li></li>
      <li></li>
    </ul>
    <p><a href="http://www.ic.unicamp.br/"><img src="img/logo-ic-40_anos-250x298.png" width="160" height="191" alt="logoic" /></a> </p>
    <p><a href="http://www.unicamp.br/unicamp/"><img src="img/logo_unicamp.gif" width="160" height="159" alt="logounicamp" /></a></p>
    <p><

In [4]:
print(ord('H'))
print(ord('e'))
print(ord('\n'))
print(ord('\r'))
'Hi'<'zzz'

72
101
10
13


True

In [3]:
# To run this, download the BeautifulSoup zip file
# http://www.py4e.com/code3/bs4.zip
# and unzip it in the same directory as this file

from urllib.request import urlopen
from bs4 import BeautifulSoup
import ssl

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
html = urlopen(url, context=ctx).read()
soup = BeautifulSoup(html, "html.parser")

# Retrieve all of the anchor tags
tags = soup('a')
for tag in tags:
    # Look at the parts of a tag
    print('TAG:', tag)
    print('URL:', tag.get('href', None))
    print('Contents:', tag.contents[0])
    print('Attrs:', tag.attrs)

Enter - http://py4e-data.dr-chuck.net/comments_42.html


In [28]:
nums = []
table = soup.find('table')
table_data = table.find_all('span')
for data in table_data:
    nums.append(int(data.text))
print(sum(nums))

2553


In [45]:
#%%time
from urllib.request import urlopen
from bs4 import BeautifulSoup
import ssl

# ejemplo = 'http://py4e-data.dr-chuck.net/comments_42.html'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
req = Request(url, headers={'User-Agent': 'Mozilla/5.0'})
with urlopen(req) as response:
    html = response.read()
    soup = BeautifulSoup(html, 'html.parser')
    # Retrieve and sum all numbers in table
    nums = []
    table = soup.find('table')
    table_data = table.find_all('span')
    for data in table_data:
        nums.append(int(data.text))
print(sum(nums))

Enter - http://py4e-data.dr-chuck.net/comments_42.html
2553


In [46]:
%%time
from urllib.request import urlopen
from bs4 import BeautifulSoup
import ssl

# ejemplo = 'http://py4e-data.dr-chuck.net/comments_373491.html'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
req = Request(url, headers={'User-Agent': 'Mozilla/5.0'})
with urlopen(req, context=ctx) as response:
    html = response.read()
    soup = BeautifulSoup(html, 'html.parser')
    # Retrieve and sum all numbers in table
    nums = []
    table = soup.find('table')
    table_data = table.find_all('span')
    for data in table_data:
        nums.append(int(data.text))
print(sum(nums))

Enter - http://py4e-data.dr-chuck.net/comments_373491.html
2248
CPU times: user 34.1 ms, sys: 0 ns, total: 34.1 ms
Wall time: 698 ms


In [7]:
%%time
import urllib.request, urllib.parse, urllib.error
from bs4 import BeautifulSoup
import ssl

# ejemplo = 'http://py4e-data.dr-chuck.net/known_by_Fikret.html'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter URL: ')
count = int(input('Enter count: '))
position = int(input('Enter position: '))

for c in range(count+1):
    print('Retrieving: '+url)
    html = urllib.request.urlopen(url, context=ctx).read()
    soup = BeautifulSoup(html, 'html.parser')
    # Retrieve urls and names of the anchor tags in a list
    urls = []
    names = []
    tags = soup('a')
    for tag in tags:
        urls.append(tag.get('href', None))
        names.append(tag.text)
    url = urls[position-1]
print(names[position-1])

Enter URL: http://py4e-data.dr-chuck.net/known_by_Davy.html
Enter count: 7
Enter position: 18
Retrieving: http://py4e-data.dr-chuck.net/known_by_Davy.html
Retrieving: http://py4e-data.dr-chuck.net/known_by_Wyatt.html
Retrieving: http://py4e-data.dr-chuck.net/known_by_Khalan.html
Retrieving: http://py4e-data.dr-chuck.net/known_by_Alieshah.html
Retrieving: http://py4e-data.dr-chuck.net/known_by_Tanvir.html
Retrieving: http://py4e-data.dr-chuck.net/known_by_Raymond.html
Retrieving: http://py4e-data.dr-chuck.net/known_by_Elice.html
Retrieving: http://py4e-data.dr-chuck.net/known_by_Justine.html


KeyboardInterrupt: 

In [9]:
import xml.etree.ElementTree as ET

data = '''
<person>
<name>Chuck</name>
<phone type="intl">
+1 734 303 4456
</phone>
<email hide="yes" />
</person>'''

tree = ET.fromstring(data)
print('Name:', tree.find('name').text)
print('Attr:', tree.find('email').get('hide'))

Name: Chuck
Attr: yes


In [17]:
import xml.etree.ElementTree as ET

input  = '''
<stuff>
    <users>
        <user x="2">
            <id>001</id>
            <name>Chuck</name>
        </user>
        <user x="7">
            <id>009</id>
            <name>Brent</name>
        </user>
    </users>
</stuff>'''

stuff = ET.fromstring(input)
lst = stuff.findall('users/user')
print('User count:', len(lst))

for item in lst:
    print('Name', item.find('name').text)
    print('Id', item.find('id').text)
    print('Attribute', item.get('x'))

User count: 2
Name Chuck
Id 001
Attribute 2
Name Brent
Id 009
Attribute 7


In [21]:
import xml.etree.ElementTree as ET

input  = '''
<stuff>
    <users>
        <user x="2">
            <id>001</id>
            <name>Chuck</name>
        </user>
        <user x="7">
            <id>009</id>
            <name>Brent</name>
        </user>
    </users>
</stuff>'''

stuff = ET.fromstring(input)

lst = stuff.findall('users/user')
print('User count (users/user):', len(lst))

lst2 = stuff.findall('users')
print('User count (users):', len(lst2))

lst3 = stuff.findall('user')
print('User count (user):', len(lst3))

User count (users/user): 2
User count (users): 1
User count (user): 0


In [23]:
import json

data = '''
[
    { "id" : "001",
        "x" : "2",
        "name" : "Chuck"
    } ,
    { "id" : "009",
        "x" : "7",
        "name" : "Brent"
    }
]'''

info = json.loads(data)
print('User count:', len(info))

for item in info:
    print('Name', item['name'])
    print('Id', item['id'])
    print('Attribute', item['x'])

User count: 2
Name Chuck
Id 001
Attribute 2
Name Brent
Id 009
Attribute 7


In [24]:
print(info)

[{'id': '001', 'x': '2', 'name': 'Chuck'}, {'id': '009', 'x': '7', 'name': 'Brent'}]


In [5]:
from urllib.request import urlopen
import urllib.parse as par
import json
import ssl

api_key = False
# If you have a Google Places key, enter it here
# api_key = 'AIzaSy___IDByT70'
# https://developers.google.com/maps/documentation/geocoding/intro

if api_key is False:
    api_key = 42
    serviceurl = 'http://py4e-data.dr-chuck.net/json?'
else:
    serviceurl = 'https://maps.googleapis.com/maps/api/geocode/json?'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

while True:
    address = input('Enter location: ')
    if not address:  break
    
    parms = dict()
    parms['address'] = address
    if api_key is not False: parms['key'] = api_key
    url = serviceurl + par.urlencode(parms)
    
    print('Retrieving', url)
    uh = urlopen(url, context=ctx)
    data = uh.read().decode()
    print('Retrieved', len(data), 'characters')
    
    try:
        js = json.loads(data)
    except:
        js = None
    
    if not js or 'status' not in js or js['status'] != 'OK':
        print('==== Failure To Retrieve ====')
        print(data)
        continue
    
    print(json.dumps(js, indent=4))
    
    lat = js['results'][0]['geometry']['location']['lat']
    lng = js['results'][0]['geometry']['location']['lng']
    print('lat', lat, 'lng', lng)
    location = js['results'][0]['formatted_address']
    print(location)
    try:
        country_code = js['results'][0]['address_components'][3]['short_name']
    except:
        print('Not have country code')
    else:
        print(country_code)

Enter location: Atlantic Ocean
Retrieving http://py4e-data.dr-chuck.net/json?address=Atlantic+Ocean&key=42
Retrieved 1228 characters
{
    "results": [
        {
            "address_components": [
                {
                    "long_name": "Atlantic Ocean",
                    "short_name": "Atlantic Ocean",
                    "types": [
                        "establishment",
                        "natural_feature"
                    ]
                }
            ],
            "formatted_address": "Atlantic Ocean",
            "geometry": {
                "bounds": {
                    "northeast": {
                        "lat": 68.1962901,
                        "lng": 19.9970779
                    },
                    "southwest": {
                        "lat": -75.9725902,
                        "lng": -82.9931606
                    }
                },
                "location": {
                    "lat": -14.5994134,
                    "lng": -28.

In [6]:
%%time
from urllib.request import urlopen, Request
import urllib.parse, urllib.error
import xml.etree.ElementTree as ET
import ssl

# ejemplo = 'http://py4e-data.dr-chuck.net/comments_42.xml'
# assigment = 'http://py4e-data.dr-chuck.net/comments_373493.xml'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
req = Request(url, headers={'User-Agent': 'Mozilla/5.0'})
with urlopen(req, context=ctx) as response:
    html = response.read()
    print('Retrieved', len(html), 'characters')
    print(html.decode())
    tree = ET.fromstring(html)
    lst = tree.findall('comments/comment')
    print('Comments count:', len(lst))

    suma = 0
    for item in lst:
        a = int(item.find('count').text)
        suma += a
print(suma)

Enter - http://py4e-data.dr-chuck.net/comments_373493.xml
Retrieved 4249 characters
<?xml version="1.0" encoding="UTF-8"?>
<commentinfo>
  <note>This file contains the actual data for your assignment - good luck!</note>

  <comments>
    <comment>
       <name>Nicholas</name>
       <count>97</count>
    </comment>
    <comment>
       <name>Sweet</name>
       <count>96</count>
    </comment>
    <comment>
       <name>Estelle</name>
       <count>93</count>
    </comment>
    <comment>
       <name>Tareena</name>
       <count>93</count>
    </comment>
    <comment>
       <name>Limbiadhe</name>
       <count>92</count>
    </comment>
    <comment>
       <name>Felicia</name>
       <count>89</count>
    </comment>
    <comment>
       <name>Jamee</name>
       <count>88</count>
    </comment>
    <comment>
       <name>Michal</name>
       <count>85</count>
    </comment>
    <comment>
       <name>Erik</name>
       <count>85</count>
    </comment>
    <comment>
       <name>Tanies

In [4]:
%%time
from urllib.request import urlopen, Request
import urllib.parse, urllib.error
import json
import ssl

# ejemplo = 'http://py4e-data.dr-chuck.net/comments_42.json'
# assigment = 'http://py4e-data.dr-chuck.net/comments_373494.json'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter location:')
req = Request(url, headers={'User-Agent': 'Mozilla/5.0'})
print('Retrieving ',url)
with urlopen(req, context=ctx) as response:
    html = response.read().decode()
    print('Retrieved', len(html), 'characters')
    try:
        comments = json.loads(html)
    except:
        comments = None
    if not comments:
        print('==== Failure To Retrieve ====')
        print(html)
    suma = 0
    for comment in comments['comments']:
        suma += comment['count']
    print('Count: ',len(comments['comments']))
print('Sum: ',suma)

Enter location:http://py4e-data.dr-chuck.net/comments_373494.json
Retrieving  http://py4e-data.dr-chuck.net/comments_373494.json
Retrieved 2727 characters
Count:  50
Sum:  2661
CPU times: user 35 ms, sys: 472 µs, total: 35.5 ms
Wall time: 4min 23s


In [7]:
import urllib.request, urllib.parse, urllib.error
import json
import ssl

api_key = 42
serviceurl = 'http://py4e-data.dr-chuck.net/json?'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

address = input('Enter location: ')

parms = dict()
parms['address'] = address
parms['key'] = api_key
url = serviceurl + urllib.parse.urlencode(parms)

print('Retrieving', url)
uh = urllib.request.urlopen(url, context=ctx)
data = uh.read().decode()
print('Retrieved', len(data), 'characters')

try:
    js = json.loads(data)
except:
    js = None

if not js or 'status' not in js or js['status'] != 'OK':
    print('==== Failure To Retrieve ====')
    print(data)
    
if js['results'][0]['place_id']:
    print('Place id ', js['results'][0]['place_id'])
#print(json.dumps(js, indent=4))

Enter location: Universidad de Los Andes Colombia
Retrieving http://py4e-data.dr-chuck.net/json?address=Universidad+de+Los+Andes+Colombia&key=42
Retrieved 2091 characters
Place id  ChIJNbMkoaSZP44Ry7kdPWj06Qg
